design/aws-ecs/aws-ecs-service-addons.md

```diff
@@ -0,0 +1,108 @@
+
+A combination of these hooks should be able to implement any type of functionality that needs to be attached to an ECS service. If we identify new use cases that aren't served by this collection of hooks we can add new hooks to enable new extension types without breaking existing extensions.
+
+## Environments
+
+As part of this proposal there is also a new construct called `Environment`. This construct is designed to be similar to AWS Copilot environments. The intention is that an environment is a transparent wrapper for a VPC, cluster, and any other resources that the environment needs. The environment understands whether it is using EC2 or Fargate capacity. A `Service` is added to an `Environment`. The `Service` changes to EC2 mode if the environment is set to have EC2 capacity. The `Service` changes to Fargate made if the environment is set to have Fargate capacity.
+
+There is potential here for customers to eventually add an extension to an environment as well as adding extensions to a service. Adding an extension to the environment would automatically enable that extension on all services that are added to the environment. This would allow operation's minded orgs to enforce for example that all services in an environment have a mandatory security sidecar attached. It would also serve as a simplification for attaching an extension globally if there is an extension that needs to be on every single service anyway.
+
+## Future Looking
+
+The service class already supports extensions. A future goal is to add support for a new feature called "traits". The idea behind this is to allow customers to modify the behavior of all of their extensions via higher level intentions about how the service should behave. Traits are higher level adjectives that describe how the service should be configured. For example some traits could be:
+
+- "cost optimized"
+- "performance optimized"
+- "high availability"
+
+The service would then self configure based on the traits that the customer selected. For example if the service had the "cost optimized" trait for a development environment it might choose to deploy the smallest possible AWS Fargate task size, and only one of them. However the "performance optimized" or "high availability" traits would override this to deploy multiple copies that are larger.
+
+Each extension would be responsible for implementing traits however it choses. For example the FireLens logging extension might decide to implement "cost optimized" by setting a TTL on log files so that they don't accumulate and cost money for storage. However if there was a "data retention" trait that would instead cause FireLens to retain all logs forever for auditing.
+
+This "traits" feature needs more thought before implementation, but it has potential to make it much easier for customer's to define their intention for behavior without needing to explicitly set lower level settings.
```

Comment:
I assume this is out of scope for the current design?

Reply:
Yeah I had it as a stretch goal, but I think given the broad scope of what is already in the PR it would be best to move this stretch goal to its own future enhancement once we get this out in preview already for customers to play with and give feedback